Initial fix for #969

KaiKorla opened the pull request:
I tried fixing #969. I'm ok with the sidebar context menu, but unsure with the buffer context menu. Please take a look and give me some feedback. tyvm.

KaiKorla committed:
Initial fix for #969

KaiKorla commented:
Discussion @ #halloy:

x offset + 5

<img width="237" alt="image" src="https://github.com/user-attachments/assets/82271936-d1c5-451a-88ee-99d09f3a978a" />

<img width="237" alt="image" src="https://github.com/user-attachments/assets/6bf8ce24-ba0e-4c16-af77-05b79f65653d" />

<img width="151" alt="image" src="https://github.com/user-attachments/assets/a641e04b-a5f4-4c5f-8aa0-f2e52cfc5975" />

casperstorm commented:
> Discussion @ #halloy:
> 
> x offset + 5
> <img alt="image" width="237" src="https://private-user-images.githubusercontent.com/107122965/440021906-82271936-d1c5-451a-88ee-99d09f3a978a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDYyMjIzMTcsIm5iZiI6MTc0NjIyMjAxNywicGF0aCI6Ii8xMDcxMjI5NjUvNDQwMDIxOTA2LTgyMjcxOTM2LWQxYzUtNDUxYS04OGVlLTk5ZDA5ZjNhOTc4YS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNTAyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDUwMlQyMTQwMTdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kOTE3YjUwN2Y3ZDgzNjQyNzExNGZiZTJhZDZjZTg0ZTdhYmQwZjg0YzkxYTY5MDE0ZDI3ZGM0OGFhYjRhNzI4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.voYkuf3yIFiIvYnK-sCBNvSV9OjkMIRxPz7TcMO3big"> <img alt="image" width="237" src="https://private-user-images.githubusercontent.com/107122965/440022145-6bf8ce24-ba0e-4c16-af77-05b79f65653d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDYyMjIzMTcsIm5iZiI6MTc0NjIyMjAxNywicGF0aCI6Ii8xMDcxMjI5NjUvNDQwMDIyMTQ1LTZiZjhjZTI0LWJhMGUtNGMxNi1hZjc3LTA1Yjc5ZjY1NjUzZC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNTAyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDUwMlQyMTQwMTdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jNGViODA2YzkwYjU2MmMxZDFkNTc4NmNjM2M3OGYyOTQxMjRmZjJjZjRlZjMxNTNmZWExMTVlMWM3ODg2NWYyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.JXcQDxvDEmek5i8DuZZh3M4MfoOvwnbzqFYXfm40W6M"> <img alt="image" width="151" src="https://private-user-images.githubusercontent.com/107122965/440022449-a641e04b-a5f4-4c5f-8aa0-f2e52cfc5975.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDYyMjIzMTcsIm5iZiI6MTc0NjIyMjAxNywicGF0aCI6Ii8xMDcxMjI5NjUvNDQwMDIyNDQ5LWE2NDFlMDRiLWE1ZjQtNGM1Zi04YWEwLWYyZTUyY2ZjNTk3NS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNTAyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDUwMlQyMTQwMTdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hOWZmYjg0Y2Q1ZTIwNTgzZWU1ODNkMWE5ZDVjNzQ1ZTAzMjRiZmU0YWI4YmRhYTViNjRlMGVjMmUxNjVhNDkxJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.UOwEc9IUvJ1a8yloxxV4gvaaerN2gq141B5WX8GFwAU">

I think that looks good.

KaiKorla committed:
Add 5 points offset

KaiKorla committed:
- implement an option to anchor context_menu to a widget
- Restore "old" behavior (anchor on cursor) with 5.0 points offset for buffer and user context_menu
- Using "new" behavior (anchor to element) with 5.0 points offset for the sidebar context_menu

casperstorm commented:
Sorry for late review on this, @KaiKorla.
This is good, however a small new problem. When you click the sidebar-button it opens and looks great, however if you then instantly press the button again it actually doesn't close it but close then re-opens. Which is techinically what we wanted to fix initially :D

KaiKorla committed:
- introducing ContextMenuAnchor to anchoring the context menu to the mouse cursor or the widget itself
- introducing ContextMenuReclickMode to give the choice if the context menu will close on reclick or kept open
- reimplement the flow for update() to handle mouse events, actions and next status in clear way

casperstorm committed:
Renamed

casperstorm commented:
I just pushed a commit with some name changes, hope that was ok.
Last issue you could address, if you right click a user in a message, it will never close that context menu.

KaiKorla committed:
Closing context menu, when a menu entry was clicked.

KaiKorla commented:
It's working so far, but the problem with "inline" nicknames in the buffer consists. I tried to track it down, but with my limited experience in Rust and iced it is not so easy.
IMO the problem with this is, that the state cannot be found for this text-parts and therefore the implementation in my code get always a ```Status::Closed``` on ```tree.state.downcast_mut::<State>();```. I took a look into [selectable_rich_text.rs](url) and wonder about the handling of context_menu there. I don't know if this will cause problems. Maybe i'm totally wrong, but could need some advice anyhow.

@tarkah: @casperstorm told me i should mention you for any advice/help/ideas. so if you got some time please take a look. thank you very much.

casperstorm commented:
This is not forgotten, but we will need @tarkah to chime in here. Just FYI @KaiKorla.

andymandias commented on src/widget/context_menu.rs:
```diff
@@ -177,43 +197,67 @@ where
+                (true, Status::Closed, ToggleBehavior::Close, Some(position))
+                => Status::Open(position),
+
+                (false, Status::Open(_), _, None)
+                |
+                (true, Status::Open(_), _, None)
+                |
+                (true, Status::Open(_), ToggleBehavior::Close, Some(_))
+                => Status::Closed,
```
We'd want to include `(_, Status::Closed(true), _, _)` in this arm, and have `next_status` set to `Status::Closed(false)`.

andymandias commented on src/widget/context_menu.rs:
```diff
@@ -510,24 +554,6 @@ where
         clipboard: &mut dyn Clipboard,
         shell: &mut Shell<'_, Message>,
     ) {
-        if let Event::Mouse(mouse::Event::ButtonPressed(_)) = &event {
-            if cursor.position_over(layout.bounds()).is_none() {
-                self.state.status = Status::Closed;
-            }
-        }
-
-        if let Event::Mouse(mouse::Event::ButtonReleased(mouse::Button::Left)) =
-            &event
-        {
-            if cursor.position_over(layout.bounds()).is_some() {
-                self.state.status = Status::Closed;
-            }
-        }
-
-        if matches!(self.state.status, Status::Closed) {
-            shell.request_redraw();
-        }
-
```
I believe you removed this in order to avoid closing the context menu as part of the click, only to have the underlying widget re-open the menu (as it now sees it as closed).  As you discovered, the way `selectable_rich_text.rs` utilizes `context_menu` expects this behavior.  To avoid messing with `selectable_rich_text.rs`, which we would want to bring in a consult for, we can instead keep this code add a `bool` to `Status::Closed` like:
```rust
pub enum Status {
    Closed(bool),
    Open(Point),
}
```
which can be initialized as `Status::Closed(false)`.  Then, this code can become
```rust
        if let Event::Mouse(mouse::Event::ButtonPressed(_)) = &event {
            if cursor.position_over(layout.bounds()).is_none() {
                self.state.status = Status::Closed(true);
            }
        }

        if let Event::Mouse(mouse::Event::ButtonReleased(mouse::Button::Left)) =
            &event
        {
            if cursor.position_over(layout.bounds()).is_some() {
                self.state.status = Status::Closed(true);
            }
        }
```

andymandias commented on src/widget/context_menu.rs:
```diff
@@ -177,43 +197,67 @@ where
+                self.toggle_behavior,
+                position,
+            ) {
+                (true, _, ToggleBehavior::KeepOpen, Some(position))
+                |
+                (true, Status::Closed, ToggleBehavior::Close, Some(position))
+                => Status::Open(position),
```
We'd want to have the second expression in this arm only match against `Status::Closed(false)`, preventing the close→immediately-reopen issue.

andymandias commented on src/widget/context_menu.rs:
```diff
@@ -177,43 +197,67 @@ where
+            };
+
+            if next_status != prev_status {
+                state.status = next_status;
```
We should probably check that `!matches!(next_status, Status::Closed(_)) || !matches!(prev_status, Status::Closed(_))` before requesting a redraw.

andymandias reviewed (commented):
Apologies for the delay.  I think we can make a few changes that will avoid touching the more complicated bits of the code, and make this PR easier to land.  I've given the rough outline for the strategy in code comments.  Let me know if further details would be helpful.

andymandias committed:
Add boolean to to `context_menu::Status::Closed` to indicate to subsequent `update` calls whether the context menu was just closed or not.

andymandias committed:
Add thanks.

andymandias commented:
Helping get this over the line a bit, thank you for getting this most of the way complete @KaiKorla.  Works as intended on Linux for me now.  @casperstorm, when you have a chance could you confirm it's working on Mac/Windows as well?

casperstorm reviewed (approved):
This works great. Confirmed working on my end.
Thanks both @andymandias and @KaiKorla for this!
